client/allocrunner/taskrunner/vault_hook.go

```diff
@@ -311,23 +328,76 @@ OUTER:
 
 // deriveVaultToken derives the Vault token using exponential backoffs. It
 // returns the Vault token and whether the manager should exit.
-func (h *vaultHook) deriveVaultToken() (token string, exit bool) {
+func (h *vaultHook) deriveVaultToken() (string, bool) {
 	var attempts uint64
 	var backoff time.Duration
+	var err error
+	var exit bool
+	var token string
+
 	for {
-		tokens, err := h.client.DeriveToken(h.alloc, []string{h.taskName})
-		if err == nil {
-			return tokens[h.taskName], false
+		switch h.authMethod {
+		case vaultAuthMethodToken:
+			token, exit, err = h.deriveVaultTokenLegacy()
+		case vaultAuthMethodJWT:
+			token, exit, err = h.deriveVaultTokenJWT()
+		}
+		if err == nil || exit {
+			return token, exit
+		}
+
+		// Handle the retry case
+		backoff = helper.Backoff(vaultBackoffBaseline, vaultBackoffLimit, attempts)
+		attempts++
+
+		h.logger.Error("failed to derive Vault token", "error", err, "recoverable", true, "backoff", backoff)
+
+		// Wait till retrying
+		select {
+		case <-h.ctx.Done():
+			return "", true
+		case <-time.After(backoff):
 		}
+	}
+}
 
+func (h *vaultHook) deriveVaultTokenJWT() (string, bool, error) {
+	// Wait for signed identity.
+	signed, err := h.widmgr.Get(cstructs.TaskIdentity{
+		TaskName:     h.taskName,
+		IdentityName: "vault",
+	})
```

Comment:
Just checking my understanding here: we don't need to use `Watch` because we don't actually care about refreshed identities until we have to derive a brand new Vault token rather than renewing the Vault token, right?

Reply:
Yes, but also `Watch` can only be called once per identity, and the `identity` hook is already watching it. I'm OK either way, but to use `Watch` here we need to fix this:
https://github.com/hashicorp/nomad/pull/18431#discussion_r1330781301

Reply:
As per internal discussions, `Get` was kept in #18431, so I will keep using it.